File: Dart/web/Model/State.dart
import 'dart:html';

import 'Database.dart';
import 'GUI.dart';
import 'Settings.dart';
import 'TList.dart';
import 'TStep.dart';
import 'Todo.dart';

Function GetTodayTodos = (TList tlist, Function callback) {
  //Database.Current().GetTodosByParentId(tlist.Id, (todos) {
  //  callback(todos);
  //});
};

Function AddTodayTodos = (String text, TList tlist, Function callback) {
  //Database.Current().AddTodo(text, tlist.Id, callback);
};

Function GetImportantTodos = (TList tlist, Function callback) {
  //Database.Current().GetTodosByParentId(tlist.Id, (todos) {
  //  callback(todos);
  //});
};

Function AddImportantTodos = (String text, TList tlist, Function callback) {
  //Database.Current().AddTodo(text, tlist.Id, callback);
};

Function GetPlannedTodos = (TList tlist, Function callback) {
  //Database.Current().GetTodosByParentId(tlist.Id, (todos) {
  //  callback(todos);
  //});
};

class State {
  static final State Current = State._state();

  factory State() {
    return Current;
  }

  State._state();

  TList TodayTodos = TList(
      id: -1,
      name: 'Dnešní úkoly',
      icon: 'u-text-red bi bi-star',
      unfinishedOpened: true,
      finishedOpened: false,
      orderTodo: eOrderTodo.None);
  TList ImportantTodos = TList(
      id: -1,
      name: 'Důležité úkoly',
      icon: 'u-text-yellow bi bi-sun',
      unfinishedOpened: true,
      finishedOpened: false,
      orderTodo: eOrderTodo.None);
  TList PlannedTodos = TList(
      id: -1,
      name: 'Plánované úkoly',
      icon: 'u-text-blue bi bi-calendar',
      unfinishedOpened: true,
      finishedOpened: false,
      orderTodo: eOrderTodo.None);
  TList UndividedTodos = TList(
      id: -1,
      name: 'Nezařazené úkoly',
      icon: 'u-text-green bi bi-check-all',
      unfinishedOpened: true,
      finishedOpened: false,
      orderTodo: eOrderTodo.None);
  List<TList> TLists = [];
  List<Todo> Todos = [];
  List<TStep> TSteps = [];
  TList CurrentTList;
  Todo CurrentTodo;
  List<Todo> UnfinishedTodos = [];
  List<Todo> FinishedTodos = [];

  void AppInitialize() {
    TodayTodos.GetTodos = GetTodayTodos;
    TodayTodos.AddTodo = AddTodayTodos;
    ImportantTodos.GetTodos = GetImportantTodos;
    ImportantTodos.AddTodo = AddImportantTodos;
    PlannedTodos.GetTodos = GetPlannedTodos;

    Database.Current.InitializeDatabase(callback: () {
      GUI.Current.Initialize();
      LoadTLists(callback: () {
        SwitchTList(TodayTodos);
      });
    });
  }

  void ImportData(e) {
    try {
      //var reader = new FileReader();
      //reader.onload = (event) {
      //    Database.Current.ImportData(event.target.result, callback: () {
      //        LoadTLists(callback: () {
      //            SwitchTList(TodayTodos);
      //        });
      //    });
      //};
      //reader.readAsText(event.target.files[0]);
    } catch (err) {
      GUI.Current.ShowTextDialog('Chyba', 'Nastala chyba při čtení souboru');
    }
  }

  void ExportData() {
    Database.Current.ExportData(callback: (String json) {
      if (json.isNotEmpty) {
        //let base64:string = window.btoa(json);
        //GUI.Current.DownloadFile(base64);
      } else {
        GUI.Current.ShowTextDialog(
            'Chyba', 'Nelze uložit soubor, pokud je databáze prázdná');
      }
    });
  }

  void ShowSearchView() {
    GUI.Current.ShowSearchView();
    CurrentTList = null;
    Todos = [];
    HideRightSidebar();
    GUI.Current.DrawMain();
    LoadTodos();
  }

  void Search() {
    LoadTodos();
  }

  void HideRightSidebar() {
    if (GUI.Current.RightSidebarStatus == eSidebarState.Opened) {
      GUI.Current.ToggleRightSidebar();
    }
    CurrentTodo = null;
  }

  void SetColorMode(e) {
    Settings.Current.SetDarkMode(e.checked);
    GUI.Current.SetColorMode();
  }

  void LoadTLists({callback}) {
    Database.Current.GetAllTLists(callback: (List<TList> tlist) {
      TLists = tlist;
      GUI.Current.DrawTLists();
      if (callback != null) callback();
    });
  }

  void SwitchTList(todoList, {callback}) {
    GUI.Current.HideSearchView();
    CurrentTList = todoList;
    HideRightSidebar();
    LoadTodos(callback: callback);
  }

  void AddNewTListTextKeyUp(KeyEvent e) {
    if (e.key == 'Enter') AddNewTList();
  }

  void AddNewTList() {
    var name = GUI.Current.GetNewTListName();
    if (name != null && name.isNotEmpty) {
      Database.Current.AddTlist(name, callback: () {
        GUI.Current.ClearNewTListName();
        LoadTLists();
      });
    } else {
      GUI.Current.ShowTextDialog('Chyba', 'Název seznamu nesmí být prázdný');
    }
  }

  void TListNameChange(e) {
    String text = e.value;
    if (text != null && text.isNotEmpty) {
      CurrentTList.Name = text;
      CurrentTList.Update(() {
        LoadTLists();
        GUI.Current.SetHeader();
      });
    } else {
      GUI.Current.ShowTextDialog('Chyba', 'Název seznamu nesmí být prázdný');
    }
  }

  void TListOrderTodoChange(e) {
    CurrentTList.OrderTodo = (e.value);
    CurrentTList.Update(() {
      LoadTLists(callback: () {
        LoadTodos();
      });
    });
  }

  void TListIconChange(e) {
    CurrentTList.Icon = e.value;
    CurrentTList.Update(() {
      LoadTLists();
      GUI.Current.SetHeader();
    });
  }

  void RemoveTList() {
    CurrentTList.Remove(() {
      LoadTLists(callback: () {
        SwitchTList(TodayTodos);
      });
    });
  }

  void TodoDividerClick(bool unfinishedButton) {
    if (!GUI.Current.SearchView) {
      if (unfinishedButton) {
        CurrentTList.UnfinishedOpened = !CurrentTList.UnfinishedOpened;
      } else {
        CurrentTList.FinishedOpened = !CurrentTList.FinishedOpened;
      }
      if (CurrentTList.Id != UndividedTListId) {
        CurrentTList.Update(() {
          LoadTLists(callback: () {
            GUI.Current.DrawMain();
          });
        });
      } else {
        GUI.Current.DrawMain();
      }
    }
  }

  void LoadTodos({callback}) {
    if (!GUI.Current.SearchView) {
      CurrentTList.GetTodosByOrder((List<Todo> todos) {
        Todos = todos;
        UnfinishedTodos = Todos.where((Todo todo) {
          return !todo.Done;
        });
        FinishedTodos = Todos.where((Todo todo) {
          return todo.Done;
        });
        GUI.Current.DrawTLists();
        GUI.Current.DrawMain();
        if (callback != null) callback();
      });
    } else {
      var text = GUI.Current.GetSearchText();
      if (text != null && text.isNotEmpty) {
        Database.Current.FindTodo(text, callback: (List<Todo> todos) {
          Todos = todos;
          UnfinishedTodos = Todos.where((Todo todo) {
            return !todo.Done;
          });
          FinishedTodos = Todos.where((Todo todo) {
            return todo.Done;
          });
          GUI.Current.DrawTLists();
          GUI.Current.DrawMain();
          if (callback != null) callback();
        });
      } else {
        Todos = [];
        UnfinishedTodos = [];
        FinishedTodos = [];
        GUI.Current.DrawTLists();
        GUI.Current.DrawMain();
      }
    }
  }

  void TodoSelected(int id) {
    List<Todo> todoArray = Todos.where((Todo item) {
      return item.Id == id;
    });
    var todo = todoArray[0];
    if (CurrentTodo != todo) {
      CurrentTodo = todo;
      LoadTSteps(callback: (List<TStep> steps) {
        GUI.Current.SetRightSidebar();
        if (GUI.Current.RightSidebarStatus == eSidebarState.Closed) {
          GUI.Current.ToggleRightSidebar();
        }
      });
    } else {
      HideRightSidebar();
    }
  }

  void AddNewTodoTextKeyUp(KeyEvent e) {
    if (e.key == 'Enter') AddTodo();
  }

  void AddTodo() {
    var text = GUI.Current.GetNewTodoText();
    if (text != null && text.isNotEmpty) {
      CurrentTList.AddTodo(text, CurrentTList, callback: () {
        GUI.Current.ClearNewTodoText();
        LoadTodos();
      });
    } else {
      GUI.Current.ShowTextDialog('Chyba', 'Text úkolu nesmí být prázdný');
    }
  }

  void TodoDoneChanged(int id) {
    List<Todo> todoArray = Todos.where((Todo item) {
      return item.Id == id;
    });
    var todo = todoArray[0];
    todo.Done = !todo.Done;
    todo.Update(() {
      LoadTodos(callback: () {
        if (todo == CurrentTodo) GUI.Current.SetRightSidebar();
      });
    });
  }

  void TodoImportantChanged(int id) {
    List<Todo> todoArray = Todos.where((Todo item) {
      return item.Id == id;
    });
    var todo = todoArray[0];
    todo.Important = !todo.Important;
    todo.Update(() {
      LoadTodos(callback: () {
        if (todo == CurrentTodo) GUI.Current.SetRightSidebar();
      });
    });
  }

  void TodoTextChanged(e) {
    String text = e.value;
    if (text != null && text.isNotEmpty) {
      CurrentTodo.Text = text;
      CurrentTodo.Update(() {
        LoadTodos();
      });
    } else {
      GUI.Current.ShowTextDialog('Chyba', 'Text úkolu nesmí být prázdný');
    }
  }

  void TodoEndDateChanged(e) {
    CurrentTodo.EndDate = DateTime(e.value).toString();
    CurrentTodo.Update(() {
      LoadTodos();
    });
  }

  void TodoNoteChanged(e) {
    CurrentTodo.Note = e.value;
    CurrentTodo.Update(() {
      LoadTodos();
    });
  }

  void RemoveTodo() {
    CurrentTodo.Remove(() {
      HideRightSidebar();
      LoadTodos();
    });
  }

  void LoadTSteps({callback}) {
    CurrentTodo.GetTodoSteps((List<TStep> steps) {
      TSteps = steps;
      GUI.Current.DrawTSteps();
      if (callback != null) callback();
    });
  }

  void AddNewTStepTextKeyUp(KeyEvent e) {
    if (e.key == 'Enter') AddTStep();
  }

  void AddTStep() {
    var text = GUI.Current.GetNewTStepText();
    if (text != null && text.isNotEmpty) {
      CurrentTodo.AddStep(text, () {
        GUI.Current.ClearNewTStepText();
        LoadTSteps();
      });
    } else {
      GUI.Current.ShowTextDialog('Chyba', 'Text kroku nesmí být prázdný');
    }
  }

  void TStepCheckedChanged(int id) {
    List<TStep> stepsArray = TSteps.where((TStep item) {
      return item.Id == id;
    });
    var step = stepsArray[0];
    step.Done = !step.Done;
    step.Update(() {
      LoadTSteps();
    });
  }

  void TStepTextChanged(e, int id) {
    var text = e.value;
    if (text != null && text.length > 0) {
      List<TStep> stepsArray = TSteps.where((TStep item) {
        return item.Id == id;
      });
      var step = stepsArray[0];
      step.Text = text;
      step.Update(() {
        LoadTSteps();
      });
    } else {
      GUI.Current.ShowTextDialog('Chyba', 'Text kroku nesmí být prázdný');
    }
  }

  void RemoveTStep(int id) {
    List<TStep> stepsArray = TSteps.where((TStep item) {
      return item.Id == id;
    });
    var step = stepsArray[0];
    step.Remove(() {
      LoadTSteps();
    });
  }
}

File: TypeScript/TSTodo/wwwroot/Model/State.ts
const GetTodayTodos = (tlist:TList, callback) => {
    Database.Current().GetAllTodos((results) => {
        let today: Todo[] = results.filter((item) => { return item.EndDate == new Date().toDateString(); });
        callback(today);
    });
}

const AddTodayTodos = (text:string, tlist: TList, callback) => {
    let todo: Todo = new Todo(-1, text, -1, false, false, new Date().toDateString(), new Date().toDateString(), "");
    Database.Current().AddTodoClass(todo, callback);
}

const GetImportantTodos = (tlist: TList, callback) => {
    Database.Current().GetAllTodos((results) => {
        let today: Todo[] = results.filter((item) => { return item.Important; });
        callback(today);
    });
}

const AddImportantTodos = (text: string, tlist: TList, callback) => {
    let todo: Todo = new Todo(-1, text, -1, false, true, new Date().toDateString(), DateMaxValue.toDateString(), "");
    Database.Current().AddTodoClass(todo, callback);
}

const GetPlannedTodos = (tlist: TList, callback) => {
    Database.Current().GetAllTodos((results) => {
        let today: Todo[] = results.filter((item) => { return item.EndDate != DateMaxValue.toDateString(); });
        callback(today);
    });
}

class State {
    private static state: State;
    private constructor() { }

    public static Current(): State {
        if (!this.state) {
            this.state = new State();
        }
        return this.state;
    }

    public TodayTodos: TList = new TList(-1, "Dnešní úkoly", "u-text-red bi bi-star", true, false, OrderTodo.None, GetTodayTodos, AddTodayTodos);
    public ImportantTodos: TList = new TList(-1, "Důležité úkoly", "u-text-yellow bi bi-sun", true, false, OrderTodo.None, GetImportantTodos, AddImportantTodos);
    public PlannedTodos: TList = new TList(-1, "Plánované úkoly", "u-text-blue bi bi-calendar", true, false, OrderTodo.None, GetPlannedTodos, AddTodayTodos);
    public UndividedTodos: TList = new TList(-1, "Nezařazené úkoly", "u-text-green bi bi-check-all", true, false, OrderTodo.None);
    public TLists: TList[] = [];
    public Todos: Todo[] = [];
    public TSteps: TStep[] = [];
    public CurrentTList: TList;
    public CurrentTodo: Todo;
    public UnfinishedTodos: Todo[] = [];
    public FinishedTodos: Todo[] = [];

    public AppInitialize(): void {
        Database.Current().InitializeDatabase(() => {
            GUI.Current().Initialize()
            this.LoadTLists(() => {
                this.SwitchTList(this.TodayTodos);
            });
        });
    }

    public ImportData(e): void {
        try {
            let reader = new FileReader();
            reader.onload = (event) => {
                Database.Current().ImportData(event.target.result, () => {
                    this.LoadTLists(() => {
                        this.SwitchTList(this.TodayTodos);
                    });
                });
            };
            reader.readAsText(event.target.files[0]);
        }
        catch (err) {
            GUI.Current().ShowTextDialog("Chyba", "Nastala chyba při čtení souboru");
        }
    }

    public ExportData(): void {
        Database.Current().ExportData((json: string) => {
            if (json.length > 0) {
                let base64:string = window.btoa(json);
                GUI.Current().DownloadFile(base64);
            }
            else
                GUI.Current().ShowTextDialog("Chyba", "Nelze uložit soubor, pokud je databáze prázdná");
        });
    }

    public ShowSearchView(): void {
        GUI.Current().ShowSearchView();
        this.CurrentTList = null;
        this.Todos = [];
        this.HideRightSidebar();
        GUI.Current().DrawMain();
        this.LoadTodos();
    }

    public Search(): void {
        this.LoadTodos();
    }

    public HideRightSidebar(): void {
        if (GUI.Current().RightSidebarStatus == SidebarState.Opened)
            GUI.Current().ToggleRightSidebar();
        this.CurrentTodo = null;
    }

    public SetColorMode(e): void {
        Settings.Current().SetDarkMode(e.checked);
        GUI.Current().SetColorMode();
    }

    public LoadTLists(callback?): void {
        Database.Current().GetAllTLists((tlist:TList[]) => {
            this.TLists = tlist;
            GUI.Current().DrawTLists();
            if (callback != undefined)
                callback();
        });
    }

    public SwitchTList(todoList, callback?): void {
        GUI.Current().HideSearchView();
        this.CurrentTList = todoList;
        this.HideRightSidebar();
        this.LoadTodos(callback);
    }

    public AddNewTListTextKeyUp(e): void {
        if (e.key == "Enter")
            this.AddNewTList();
    }

    public AddNewTList(): void {
        let name:string = GUI.Current().GetNewTListName();
        if (name != undefined && name.length > 0) {
            Database.Current().AddTlist(name, () => {
                GUI.Current().ClearNewTListName();
                this.LoadTLists();
            });
        }
        else
            GUI.Current().ShowTextDialog("Chyba", "Název seznamu nesmí být prázdný");
    }

    public TListNameChange(e): void {
        let text:string = e.value;
        if (text != undefined && text.length > 0) {
            this.CurrentTList.Name = text;
            this.CurrentTList.Update(() => {
                this.LoadTLists();
                GUI.Current().SetHeader();
            });
        }
        else
            GUI.Current().ShowTextDialog("Chyba", "Název seznamu nesmí být prázdný");
    }

    public TListOrderTodoChange(e): void {
        this.CurrentTList.OrderTodo = parseInt(e.value);
        this.CurrentTList.Update(() => {
            this.LoadTLists(() => {
                this.LoadTodos();
            });
        });
    }

    public TListIconChange(e): void {
        this.CurrentTList.Icon = e.value;
        this.CurrentTList.Update(() => {
            this.LoadTLists();
            GUI.Current().SetHeader();
        });
    }

    public RemoveTList(): void {
        this.CurrentTList.Remove(() => {
            this.LoadTLists(() => {
                this.SwitchTList(this.TodayTodos);
            });
        });
    }

    public TodoDividerClick(unfinishedButton: boolean): void {
        if (!GUI.Current().SearchView) {
            if (unfinishedButton)
                this.CurrentTList.UnfinishedOpened = !this.CurrentTList.UnfinishedOpened;
            else
                this.CurrentTList.FinishedOpened = !this.CurrentTList.FinishedOpened;
            if (this.CurrentTList.Id != UndividedTListId) {
                this.CurrentTList.Update(() => {
                    this.LoadTLists(() => {
                        GUI.Current().DrawMain();
                    });
                });
            }
            else
                GUI.Current().DrawMain();
        }
    }

    public LoadTodos(callback?): void {
        if (!GUI.Current().SearchView) {
            this.CurrentTList.GetTodosByOrder((todos:Todo[]) => {
                this.Todos = todos;
                this.UnfinishedTodos = this.Todos.filter((todo: Todo) => { return !todo.Done; });
                this.FinishedTodos = this.Todos.filter((todo: Todo) => { return todo.Done; });
                GUI.Current().DrawTLists()
                GUI.Current().DrawMain()
                if (callback != undefined)
                    callback();
            });
        }
        else {
            let text: string = GUI.Current().GetSearchText();
            if (text != undefined && text.length > 0) {
                Database.Current().FindTodo(text, (todos:Todo[]) => {
                    this.Todos = todos;
                    this.UnfinishedTodos = this.Todos.filter((todo: Todo) => { return !todo.Done; });
                    this.FinishedTodos = this.Todos.filter((todo: Todo) => { return todo.Done;});
                    GUI.Current().DrawTLists();
                    GUI.Current().DrawMain();
                    if (callback != undefined)
                        callback();
                });
            }
            else {
                this.Todos = [];
                this.UnfinishedTodos = [];
                this.FinishedTodos = [];
                GUI.Current().DrawTLists();
                GUI.Current().DrawMain();
            }
        }
    }

    public TodoSelected(id:number): void {
        let todoArray: Todo[] = this.Todos.filter((item:Todo) => { return item.Id == id });
        let todo: Todo = todoArray[0];
        if (this.CurrentTodo != todo) {
            this.CurrentTodo = todo;
            this.LoadTSteps((steps:TStep[]) => {
                GUI.Current().SetRightSidebar();
                if (GUI.Current().RightSidebarStatus == SidebarState.Closed)
                    GUI.Current().ToggleRightSidebar();
            });
        }
        else
            this.HideRightSidebar();
    }

    public AddNewTodoTextKeyUp(e): void {
        if (e.key == "Enter")
            this.AddTodo();
    }

    public AddTodo(): void {
        let text:string = GUI.Current().GetNewTodoText();
        if (text != undefined && text.length > 0) {
            this.CurrentTList.AddTodo(text, this.CurrentTList, () => {
                GUI.Current().ClearNewTodoText();
                this.LoadTodos();
            });
        }
        else
            GUI.Current().ShowTextDialog("Chyba", "Text úkolu nesmí být prázdný");
    }

    public TodoDoneChanged(id:number): void {
        let todoArray:Todo[] = this.Todos.filter((item:Todo) => { return item.Id == id });
        let todo:Todo = todoArray[0];
        todo.Done = !todo.Done;
        todo.Update(() => {
            this.LoadTodos(() => {
                if (todo == this.CurrentTodo)
                    GUI.Current().SetRightSidebar();
            });
        });
    }

    public TodoImportantChanged(id: number): void {
        let todoArray: Todo[] = this.Todos.filter((item: Todo) => { return item.Id == id });
        let todo:Todo = todoArray[0];
        todo.Important = !todo.Important;
        todo.Update(() => {
            this.LoadTodos(() => {
                if (todo == this.CurrentTodo)
                    GUI.Current().SetRightSidebar();
            });
        });
    }

    public TodoTextChanged(e): void {
        let text: string = e.value;
        if (text != undefined && text.length > 0) {
            this.CurrentTodo.Text = text;
            this.CurrentTodo.Update(() => {
                this.LoadTodos();
            });
        }
        else
            GUI.Current().ShowTextDialog("Chyba", "Text úkolu nesmí být prázdný");
    }

    public TodoEndDateChanged(e): void {
        this.CurrentTodo.EndDate = new Date(e.value).toDateString();
        this.CurrentTodo.Update(() => {
            this.LoadTodos();
        });
    }

    public TodoNoteChanged(e): void {
        this.CurrentTodo.Note = e.value;
        this.CurrentTodo.Update(() => {
            this.LoadTodos();
        });
    }

    public RemoveTodo(): void {
        this.CurrentTodo.Remove(() => {
            this.HideRightSidebar();
            this.LoadTodos();
        });
    }

    public LoadTSteps(callback?): void {
        this.CurrentTodo.GetTodoSteps((steps: TStep[]) => {
            this.TSteps = steps;
            GUI.Current().DrawTSteps();
            if (callback != undefined)
                callback();
        });
    }

    public AddNewTStepTextKeyUp(e): void {
        if (e.key == "Enter")
            this.AddTStep();;
    }

    public AddTStep(): void {
        let text:string = GUI.Current().GetNewTStepText();
        if (text != undefined && text.length > 0) {
            this.CurrentTodo.AddStep(text, () => {
                GUI.Current().ClearNewTStepText();
                this.LoadTSteps();
            });
        }
        else
            GUI.Current().ShowTextDialog("Chyba", "Text kroku nesmí být prázdný");
    }

    public TStepCheckedChanged(id: number): void {
        let stepsArray: TStep[] = this.TSteps.filter((item: TStep) => { return item.Id == id });
        let step: TStep = stepsArray[0];
        step.Done = !step.Done;
        step.Update(() => {
            this.LoadTSteps();
        });
    }

    public TStepTextChanged(e, id:number): void {
        let text:string = e.value;
        if (text != undefined && text.length > 0) {
            let stepsArray: TStep[] = this.TSteps.filter((item: TStep) => { return item.Id == id });
            let step: TStep = stepsArray[0];
            step.Text = text;
            step.Update(() => {
                this.LoadTSteps();
            });
        }
        else
            GUI.Current().ShowTextDialog("Chyba", "Text kroku nesmí být prázdný");
    }

    public RemoveTStep(id:number): void {
        let stepsArray: TStep[] = this.TSteps.filter((item: TStep) => { return item.Id == id });
        let step:TStep = stepsArray[0];
        step.Remove(() => {
            this.LoadTSteps();
        });
    }
}

State.Current().AppInitialize();